diff --git a/tools/javascript.py b/tools/javascript.py
index 5a93584..44ae32d 100644
--- a/tools/javascript.py
+++ b/tools/javascript.py
@@ -1,8 +1,9 @@
 import os
+from SCons.Util import WhereIs


 def exists(env):
-    return "EM_CONFIG" in os.environ
+    return WhereIs("emcc") is not None


 def generate(env):
@@ -10,9 +11,6 @@ def generate(env):
         print("Only wasm32 supported on web. Exiting.")
         Exit()

-    if "EM_CONFIG" in os.environ:
-        env["ENV"] = os.environ
-
     env["CC"] = "emcc"
     env["CXX"] = "em++"
     env["AR"] = "emar"
@@ -39,7 +37,14 @@ def generate(env):
     env.Replace(SHLINKFLAGS="$LINKFLAGS")
     env.Replace(SHLINKFLAGS="$LINKFLAGS")

-    if env["target"] == "debug":
+    # Thread support (via SharedArrayBuffer).
+    env.Append(CPPDEFINES=["PTHREAD_NO_RENAME"])
+    env.Append(CCFLAGS=["-s", "USE_PTHREADS=1"])
+    env.Append(LINKFLAGS=["-s", "USE_PTHREADS=1"])
+    env.Append(LINKFLAGS=["-s", "PTHREAD_POOL_SIZE=8"])
+    env.Append(LINKFLAGS=["-s", "WASM_MEM_MAX=2048MB"])
+
+    if env["target"] in ["editor", "template_debug"]:
         env.Append(CCFLAGS=["-O0", "-g"])
-    elif env["target"] == "release":
+    elif env["target"] == "template_release":
         env.Append(CCFLAGS=["-O3"])
